Python/traceback.c

```diff
@@ -512,8 +516,147 @@ _Py_DisplaySourceLine(PyObject *f, PyObject *filename, int lineno, int indent, i
     return err;
 }
 
+/* AST based Traceback Specialization
+ *
+ * When displaying a new traceback line, for certain syntactical constructs
+ * (e.g a subscript, an arithmetic operation) we try to create a representation
+ * that separates the primary source of error from the rest.
+ * 
+ * Example specialization of BinOp nodes:
+ *  Traceback (most recent call last):
+ *    File "/home/isidentical/cpython/cpython/t.py", line 10, in <module>
+ *      add_values(1, 2, 'x', 3, 4)
+ *      ^^^^^^^^^^^^^^^^^^^^^^^^^^^
+ *    File "/home/isidentical/cpython/cpython/t.py", line 2, in add_values
+ *      return a + b + c + d + e
+ *             ~~~~~~^~~
+ *  TypeError: 'NoneType' object is not subscriptable
+ */
+
+#define IS_WHITESPACE(c) (((c) == ' ') || ((c) == '\t') || ((c) == '\f'))
+
+static int
+extract_anchors_from_expr(PyObject *segment, expr_ty expr, int *left_anchor, int *right_anchor)
+{
+    switch (expr->kind) {
+        case BinOp_kind: {
+            PyObject *operator = PyUnicode_Substring(segment, expr->v.BinOp.left->end_col_offset,
```

Comment:
This seems fine but I wonder if it would be better to just avoid the `PyUnicode_Substring` and just iterate starting from `expr->v.BinOp.left->end_col_offset` instead of `0` in the substring. I'd suggest trying it out and see if it leads to simpler code, if not this is good.

Reply:
Can you explain it a bit more in detail? If I understood correctly, it wouldn't be much different from the current form would it? If we decide that, we could probably just pass `const char *` as the segment instead of the `PyObject *` form, and then it would simplify the exception handling logic at all (we wouldn't need to return `-1`s at all, compared to just making 1 less invocation that might result with an error). Let me give it a shot